Tasks: general system for recognizing and executing service work

sam0x17 opened the pull request:
`polkadot-sdk` version of original tasks PR located here: https://github.com/paritytech/substrate/pull/14329

Fixes #206

## Status
- [x] Generic `Task` trait
- [x] `RuntimeTask` aggregated enum, compatible with `construct_runtime!`
- [x] Casting between `Task` and `RuntimeTask` without needing `dyn` or `Box`
- [x] Tasks Example pallet
- [x] Runtime tests for Tasks example pallet
- [x] Parsing for task-related macros
- [x] Retrofit parsing to make macros optional
- [x] Expansion for task-related macros
- [x] Adds support for args in tasks
- [x] Retrofit tasks example pallet to use macros instead of manual syntax
- [x] Weights
- [x] Cleanup
- [x] UI tests
- [x] Docs

## Target Syntax
Adapted from https://github.com/paritytech/polkadot-sdk/issues/206#issue-1865172283

```rust
// NOTE: this enum is optional and is auto-generated by the other macros if not present
#[pallet::task]
pub enum Task<T: Config> {
    AddNumberIntoTotal {
        i: u32,
    }
}

/// Some running total.
#[pallet::storage]
pub(super) type Total<T: Config<I>, I: 'static = ()> =
StorageValue<_, (u32, u32), ValueQuery>;

/// Numbers to be added into the total.
#[pallet::storage]
pub(super) type Numbers<T: Config<I>, I: 'static = ()> =
StorageMap<_, Twox64Concat, u32, u32, OptionQuery>;

#[pallet::tasks_experimental]
impl<T: Config<I>, I: 'static> Pallet<T, I> {
	/// Add a pair of numbers into the totals and remove them.
	#[pallet::task_list(Numbers::<T, I>::iter_keys())]
	#[pallet::task_condition(|i| Numbers::<T, I>::contains_key(i))]
	#[pallet::task_index(0)]
	pub fn add_number_into_total(i: u32) -> DispatchResult {
		let v = Numbers::<T, I>::take(i).ok_or(Error::<T, I>::NotFound)?;
		Total::<T, I>::mutate(|(total_keys, total_values)| {
			*total_keys += i;
			*total_values += v;
		});
		Ok(())
	}
}
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3521087

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3521117

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3523715

kianenigma commented on substrate/frame/support/src/traits/tasks.rs:
```diff
@@ -0,0 +1,74 @@
+
+/// Contains a subset of [`Task`] that can be generalized over the aggregated `RuntimeTask`
+/// enum.
+pub trait AggregatedTask: Sized + FullCodec + TypeInfo + Clone + Debug + PartialEq + Eq {
```
Why two traits are needed here? We have multiple similar traits that represent both the singular and the plural version of an operation (eg. `OnInitialize`). Why this doesn't fit into the same model?

sam0x17 replied:
There are some methods and bounds on `Task` that the runtime/aggregated types can't support. What I should do though I think to clean this up is make `Task` strictly a superset of this one so nothing is duplicated

kianenigma commented on substrate/frame/examples/tasks-example/src/lib.rs:
```diff
@@ -0,0 +1,179 @@
+	where
+		T: TypeInfo,
+	{
+		type Enumeration = sp_std::vec::IntoIter<Task<T>>;
```
Does this type need to be configurable? won't it always be `Vec`? 

Would be useful if you provide a fake example of the syntax that the end user will put under `#[pallet::task]` to clarify this.

sam0x17 commented:
btw this branch is going to get force-pushed over later today -- my version that actually has the correct rebase with `polkadot-sdk` is local right now. This one had some issues.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3595046

sam0x17 commented:
remaining compile issue:
```
the trait bound `RuntimeEvent: From<tasks_example::Event<Runtime>>` is not satisfied
```
is this because I'm not generating a `From` impl that I should be generating in my `RuntimeTask` macro expansion?

I just find it strange because I didn't need this impl on the `substrate` version of this branch.

KiChjang commented:
Did you forget to include the tasks_example pallet in your construct_runtime?

sam0x17 commented:
> Did you forget to include the tasks_example pallet in your construct_runtime?
lmao thank you yes I think this was one of the lines the patch file misses

sam0x17 commented:
> Did you forget to include the tasks_example pallet in your construct_runtime?

oh yeah and then the weird follow-up to that that I remember fixing for `substrate` but now forget how I fixed:

after adding `TasksExample: tasks_example` to `construct_runtime`, I then get this error on the `construct_runtime!`:

```
  error: expected one of: `Pallet`, `Call`, `Storage`, `Event`, `Error`, `Config`, `Origin`, `Inherent`, `ValidateUnsigned`, `FreezeReason`, `HoldReason`, `Task`, `LockId`, `SlashReason`
```

🤔

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3595587

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3595616

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3596964

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3597247

gupnik commented:
> > Did you forget to include the tasks_example pallet in your construct_runtime?
> 
> oh yeah and then the weird follow-up to that that I remember fixing for `substrate` but now forget how I fixed:
> 
> after adding `TasksExample: tasks_example` to `construct_runtime`, I then get this error on the `construct_runtime!`:
> 
> ```
>   error: expected one of: `Pallet`, `Call`, `Storage`, `Event`, `Error`, `Config`, `Origin`, `Inherent`, `ValidateUnsigned`, `FreezeReason`, `HoldReason`, `Task`, `LockId`, `SlashReason`
> ```
> 
> 🤔
> 
> (now pushed up on this branch)

@sam0x17 It seems that `task_part` becomes None if a pallet doesn't have any tasks. In that case, we probably want to return `quote::quote!()` from `tt_default_parts`.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3597380

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3608526

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3611761

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3611779

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3627621

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3627890

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3627919

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3630324

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3650135

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3650106

sam0x17 commented:
new one @KiChjang / @gupnik:

On current version of the branch, when compiling `substrate/bin/node/runtime`, I currently get:

```
     Compiling kitchensink-runtime v3.0.0-dev (/home/sam/workspace/polkadot-sdk/substrate/bin/node/runtime)
  error[E0532]: expected tuple struct or tuple variant, found type alias `TasksExample`
      --> /home/sam/workspace/polkadot-sdk/substrate/bin/node/runtime/src/lib.rs:2096:3
       |
  2096 |         TasksExample: tasks_example,
       |         ^^^^^^^^^^^^
       |
```

All I've changed is I've added proper impls for the `RuntimeTask` methods including `enumerate()`. Is there something obvious I'm missing?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3650752

kianenigma commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,106 @@
+			#( #task_variants )*
+		}
+
+		impl #scrate::traits::Task for RuntimeTask {
```
I thought this should implement the `Aggregate` trait?

sam0x17 replied:
refactored that out now it's `Task` all the way down -- was able to resolve the object safety issues by never using a trait object and just casting between the inner and outer enum

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3651712

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3662917

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3663275

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3664377

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3664793

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3664952

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3682137

sam0x17 commented on bridges/modules/grandpa/src/mock.rs:
```diff
@@ -59,6 +59,7 @@ impl frame_system::Config for TestRuntime {
 	type RuntimeOrigin = RuntimeOrigin;
 	type Nonce = u64;
 	type RuntimeCall = RuntimeCall;
+	type RuntimeTask = RuntimeTask;
```
@kianenigma is it OK that right now it requires this everywhere?

kianenigma replied:
Yes that's understandable.

kianenigma replied:
I think you should not spend too time on this for now. Hopefully https://github.com/paritytech/polkadot-sdk/issues/171 is done before you are done.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3738592

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3739501

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3739502

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3739503

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3744434

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3744435

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3744433

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3754084

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3764199

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3764219

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3764200

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3764201

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3766434

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3766435

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3766436

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3766453

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3775036

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3775037

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3777595

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3777577

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3790658

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3790659

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3790660

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3791867

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3807309

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3824814

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3838257

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3840113

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3840413

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3840529

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3854245

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3879510

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3879511

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3879529

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3879512

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3885483

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3885595

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3885596

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3885597

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3887750

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3889698

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3893971

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3893953

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3893952

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3893954

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3894366

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3976821

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3976819

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3976820

sam0x17 commented:
expansion is now working (though need to un-hard-code a few things) 🎉 

one issue that has cropped up though is we get this phantom rust analyzer error on pallets using tasks now:
![image](https://github.com/paritytech/polkadot-sdk/assets/1855021/37f56c71-9c9f-40e0-a901-a4743b4398b4)

I think it has something to do with this `Debug` impl we have to output for the generated task enum:

```rust
impl<T: Config> core::fmt::Debug for #ident<#type_use_generics> {
	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
		f.debug_struct(stringify!(#ident)).field("value", self).finish()
	}
}
```

Note that this same code, when implemented manually within the pallet, does not cause any RA error.

Will continue debugging. Don't want to introduce another one of these phantom RA errors since there are so few left in substrate now.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3977256

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3977255

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3977254

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3977273

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4009512

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4009649

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4012477

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4012509

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104074

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104479

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104494

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104493

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104492

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4104511

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4107000

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4107149

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4107564

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4108124

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4119938

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4138509

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4141270

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4142287

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4183250

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4183776

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/where-contracts-fail-and-runtimes-chains-are-needed/4464/6

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/where-contracts-fail-and-runtimes-chains-are-needed/4464/8

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4282415

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4282977

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4283647

sam0x17 commented on substrate/frame/examples/tasks/src/lib.rs:
```diff
@@ -0,0 +1,112 @@
+	where
+		T: TypeInfo,
+	{
+		// this will be auto-generated by the macros and will always be the same
```
@gupnik btw I think my intent was to auto-generate this as well

gupnik replied:
@sam0x17 This would actually not be a part of this pallet. Instead, we have a call in `frame_system` that will be responsibility to trigger any task defined throughout the runtime. Will remove this as part of the remaining cleanup.

sam0x17 replied:
gotcha, that makes more sense

sam0x17 commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,124 @@
+			let path = &decl.path;
+			let index = decl.index;
+
+			// TODO: `pallet` will probably not be needed when `Task` is generated by macro
```
also this TODO @gupnik

sam0x17 commented on substrate/frame/support/procedural/src/pallet/expand/tasks.rs:
```diff
@@ -0,0 +1,265 @@
+					.inputs
+					.iter()
+					.map(|input| match input {
+						// Todo: This should be checked in the parsing stage.
```
yes I agree

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4338501

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4338514

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4341104

sam0x17 commented on substrate/frame/examples/tasks/src/benchmarking.rs:
```diff
@@ -0,0 +1,42 @@
+	fn add_number_into_total() {
+		Numbers::<T>::insert(0, 1);
+
+		#[block]
```
Can't this just be `#[extrinsic_call]`?

gupnik replied:
`#[extrinsic_call]` actually requires the first argument to be an `origin`.

sam0x17 commented on substrate/frame/support/test/tests/pallet_ui/task_missing_condition.stderr:
```diff
@@ -0,0 +1,5 @@
+  --> tests/pallet_ui/task_missing_condition.rs:32:6
+   |
+32 |         fn foo() -> DispatchResult {
+   |            ^^^
```
this one should already be covered by the pure parsing tests and can be omitted I think

gupnik replied:
Keeping it for spans for now. Happy to remove once the discussion is final.

sam0x17 commented on substrate/frame/support/test/tests/pallet_ui/task_missing_index.stderr:
```diff
@@ -0,0 +1,5 @@
+  --> tests/pallet_ui/task_missing_index.rs:31:6
+   |
+31 |         fn foo() -> DispatchResult {
+   |            ^^^
```
this one also covered by existing parsing tests

gupnik replied:
Same as above.

sam0x17 commented on substrate/frame/support/test/tests/pallet_ui/task_missing_list.stderr:
```diff
@@ -0,0 +1,5 @@
+  --> tests/pallet_ui/task_missing_list.rs:33:6
+   |
+33 |         fn foo() -> DispatchResult {
+   |            ^^^
```
this one also covered by existing parsing tests

sam0x17 commented on substrate/frame/support/test/tests/pallet_ui/task_missing_weight.stderr:
```diff
@@ -0,0 +1,5 @@
+  --> tests/pallet_ui/task_missing_weight.rs:34:6
+   |
+34 |         fn foo() -> DispatchResult {
+   |            ^^^
```
this one also covered by existing parsing tests

gupnik replied:
Same as above.

sam0x17 reviewed (commented):
Looks good, added docs and some doc tests. A bunch of the UI tests you added @gupnik are things that are already fully tested by the parsing tests if you take a look at those in the `procedural/src/tests` directory. Benefit of using those is all of them together runs in < 1 second vs each UI test taking ~10 seconds to compile and then run.

Scenarios that involve a compile-error due to something that is post-expansion will always still need UI tests, but parsing (anything before we start expansion) should always be fully testable with simple unit tests against the Parse impl. Although we sometimes throw errors during expansion, this is an anti-pattern as expansion is supposed to be infallible.

gupnik commented:
> A bunch of the UI tests you added @gupnik are things that are already fully tested by the parsing tests

@sam0x17 Parsing tests are not able to ensure that the error spans are correct, which I believe are critical pieces for macro usability. Hence, I do not think that they can replace the UI tests yet. Rather, they can help in quick testing of the macro parsing logic. 

> this has my blessing if you remove the redundant UI tests

If you still think that these UI tests are redundant, please let me know the alternative that I can use to assert error spans.

> I can take a stab at this tomorrow if desired

That will be really helpful, thanks! I have encountered a couple of edge cases while building the UI tests, which I hope to complete by tomorrow and mark it ready for review.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4355480

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4355747

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/directly-unit-testing-parse-impls-to-avoid-unneeded-ui-tests/4676/1

sam0x17 commented:
followup on parsing errors and span locations, my general take is if we are fine-grained enough about doing parsing tests on items and their sub items individually, the problem vanishes: https://forum.polkadot.network/t/directly-unit-testing-parse-impls-to-avoid-unneeded-ui-tests/4676

gupnik commented:
bot fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4369585

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4370507

gavofyork commented:
Regaring the target syntax, if the `Task` type is present (and I see no good reason to include it since it's a) redundant and b) not done with `Call`), in the description's example, shouldn't it read:

```rust
#[pallet::task]
pub enum Task<T: Config> {
    AddNumberIntoTotal { i: u32 },
}
```

gupnik commented:
> I guess it could also be done with named fields which would be more readable but need a cleverer macro to avoid sprinkling the task name all over the place.

@gavofyork I have actually implemented the named fields in the current approach. It seems that the target syntax was not updated in the description correctly. It looks like the following:

```rust
 #[pallet::task]
pub enum Task<T: Config> {
    AddNumberIntoTotal {
        i: u32,
    }
}
```
Accordingly, we are able to use the following:
```rust
#[pallet::task_list(Numbers::<T, I>::iter_keys())]
#[pallet::task_condition(|i| Numbers::<T, I>::contains_key(i))]
```

gavofyork commented:
> Accordingly, we are able to use the following:

```rust
#[pallet::task_list(Numbers::<T, I>::iter_keys())]
```

This contains no name to instruct what field iteration should be used to inform. When there is only one field it doesn't matter much, but how does it work with two?

gupnik commented:
> This contains no name to instruct what field iteration should be used to inform. When there is only one field it doesn't matter much, but how does it work with two?

It basically expects a tuple. This is a dummy example with two fields that I have added as a UI Test:

```rust
#[pallet::tasks]
impl<T: Config> Pallet<T> {
    #[pallet::task_index(0)]
    #[pallet::task_condition(|i, j| i == 0u32 && j == 2u64)]
    #[pallet::task_list(vec![(0u32, 2u64), (2u32, 4u64)].iter())]
    #[pallet::task_weight(0.into())]
    fn foo(i: u32, j: u64) -> DispatchResult {
	    ensure!(i == 0, "i must be 0");
	    ensure!(j == 2, "j must be 2");
	    Ok(())
    }
}
```

gavofyork commented:
> > This contains no name to instruct what field iteration should be used to inform. When there is only one field it doesn't matter much, but how does it work with two?
> 
> It basically expects a tuple. This is a dummy example with two fields that I have added as a UI Test:
> 
> ```rust
> #[pallet::tasks]
> impl<T: Config> Pallet<T> {
>     #[pallet::task_index(0)]
>     #[pallet::task_condition(|i, j| i == 0u32 && j == 2u64)]
>     #[pallet::task_list(vec![(0u32, 2u64), (2u32, 4u64)].iter())]
>     #[pallet::task_weight(0.into())]
>     fn foo(i: u32, j: u64) -> DispatchResult {
> 	    ensure!(i == 0, "i must be 0");
> 	    ensure!(j == 2, "j must be 2");
> 	    Ok(())
>     }
> }
> ```

I see; this is a little unexpected since named fields are expected elsewhere in the API. That said, I don't see any easy alternative so I would probably be inclined to let it pass.

ggwpez commented on substrate/frame/examples/tasks/src/lib.rs:
```diff
@@ -0,0 +1,77 @@
+		NotFound,
+	}
+
+	#[pallet::tasks]
```
Can we maybe call this `pallet::experimental::tasks` until we stabilize it? There could be some API tweaks i imagine.

gupnik replied:
This one actually matches the specification as per the related issue. Do you think that it still needs the `experimental`?

ggwpez commented on substrate/frame/examples/tasks/src/lib.rs:
```diff
@@ -0,0 +1,77 @@
+		#[pallet::task_list(Numbers::<T>::iter_keys())]
+		#[pallet::task_condition(|i| Numbers::<T>::contains_key(i))]
+		#[pallet::task_weight(T::WeightInfo::add_number_into_total())]
+		#[pallet::task_index(0)]
```
This probably deviates too much from the original syntax, but is IMHO a bit more readable.
```suggestion
		#[pallet::task(Task {
				list: Numbers::<T>::iter_keys(),
				condition: &Numbers::<T>::contains_key,
				weight: T::WeightInfo::add_number_into_total(),
				index: 0,
		)]
```

gavofyork replied:
It's not so much that it deviates from the original syntax. Rather it deviates from pre-existing idioms, specifically for `pallet::call`:

```rust
		#[pallet::weight(Weight::zero())]
		#[pallet::call_index(0)]
		...
```

ggwpez commented on substrate/frame/examples/tasks/src/lib.rs:
```diff
@@ -0,0 +1,77 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
//! This pallet demonstrates the use of the `pallet::task` api for service work.
```

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+			let index = decl.index;
+
+			from_impls.push(quote! {
+				impl From<#path::Task<#runtime_name>> for RuntimeTask {
```
Maybe a `TryInto` for converting it back?

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+		#[derive(
+			Clone, Eq, PartialEq,
+			#scrate::__private::codec::Encode, #scrate::__private::codec::Decode,
+			// #scrate::__private::codec::MaxEncodedLen,
```
Can we make it MEL?

gavofyork replied:
Probably not. Call isn't; if you need to store it you can just use `Bounded`.

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+				}
+			});
+
+			task_variants.push(quote! {
```
I see there is an `expand_outer_enum` convenience function. Can it be used here?

sam0x17 replied:
I originally tried to and it broke stuff is my vague recollection

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+			#( #task_variants )*
+		}
+
+		impl #scrate::traits::Task for RuntimeTask {
```
Can you try if adding `#[automatically_derived]` removes the clippy warnings? Then we can omit the `#[allow(unreachable_code)]`.

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+			fn is_valid(&self) -> bool {
+				match self {
+					#(RuntimeTask::#variant_names(val) => val.is_valid(),)*
+					_ => unreachable!(#INCOMPLETE_MATCH_QED),
```
Its better to not have an `_` here but the concrete unreachable variant name like `__Ignore` of whatever we are using.

gupnik replied:
This is needed for a call when there is no Task defined in the Runtime leading to a `RuntimeTask` with no variants at all.

ggwpez commented on substrate/frame/support/procedural/src/construct_runtime/expand/task.rs:
```diff
@@ -0,0 +1,122 @@
+		#( #from_impls )*
+	};
+	// use proc_utils::*;
+	// output.pretty_print();
```
.

ggwpez commented on substrate/frame/support/procedural/src/pallet/expand/tasks.rs:
```diff
@@ -0,0 +1,257 @@
+				.iter()
+				.map(|task| {
+					let ident = &task.item.sig.ident;
+					let ident = format_ident!("{}", ident.to_string().to_class_case());
```
I think this kills the span.

sam0x17 replied:
it's getting quoted so it will get a new span anyway when this happens, this only matters for spans that we specifically pass to an `Error` AFAIK

ggwpez commented on substrate/frame/support/procedural/src/pallet/expand/tasks.rs:
```diff
@@ -0,0 +1,257 @@
+		let task_fn_idents = self
+			.tasks
+			.iter()
+			.map(|task| format_ident!("{}", &task.item.sig.ident.to_string().to_class_case()))
```
Not sure if the span is important here.

ggwpez commented on substrate/frame/support/procedural/src/pallet/expand/tasks.rs:
```diff
@@ -0,0 +1,257 @@
+			{
+				type Enumeration = #sp_std::vec::IntoIter<#enum_use>;
+
+				fn iter() -> Self::Enumeration {
```
I dont really see the advantage or using an `IntoIter` when it could just return a `Vec`.

gupnik replied:
The `Task` trait actually requires `type Enumeration: Iterator;`. Hence, we needed to use `IntoIter` here.

sam0x17 replied:
just gives a little bit more flexibility

ggwpez commented on substrate/frame/support/procedural/src/pallet/expand/tasks.rs:
```diff
@@ -0,0 +1,257 @@
+					}
+				}
+
+				fn is_valid(&self) -> bool {
```
`valid` means that the task is ready to run? *Valid* sounds more like it is well defined or something.  
Or maybe `can_run` since we use the word `run` already.

gavofyork replied:
Tasks are a bit like transactions in that they're `dispatchable` and can exist as an extrinsic. The main differences are that:
- Tasks are free to run;
- All valid tasks can be enumerated;
- It can always be cheaply determined whether a task will execute successfully.

We already use the term "valid" for transactions which are allowed to be included on-chain (e.g. `TransactionValidity`, `ValidTransation`). I don't think it's a stretch to use it to mean the same thing for tasks too.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509945

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509946

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509964

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509947

gupnik commented:
bot fmt

gupnik commented:
bot fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4604923

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4604958

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4605643

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4605798

gupnik commented:
bot fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4606033

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4606173

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4606310

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4606454

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4606591

ggwpez commented on substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs:
```diff
@@ -0,0 +1,259 @@
+			Ok(_) => panic!("Expected an `Error(..)`, but got Ok(..)"),
+			Err(e) => {
+				let error_message = e.to_string();
+				let re = regex::Regex::new($reg).expect("Invalid regex pattern");
```
I think it needs to use a private export of the macro here.

ggwpez commented on substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs:
```diff
@@ -0,0 +1,259 @@
+/// Although this forces all tests that use [`simulate_manifest_dir`] to run sequentially with
+/// respect to each other, this is still several orders of magnitude faster than using UI
+/// tests, even if they are run in parallel.
+static MANIFEST_DIR_LOCK: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
```
We could also fork every test into its own process, but I guess this is easier for now.

ggwpez replied:
```suggestion
static MANIFEST_DIR_LOCK: Mutex<()> = Mutex::new(());
```

ggwpez commented on substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs:
```diff
@@ -0,0 +1,259 @@
+/// Although this is primarily intended to be used with parsing errors, this macro is general
+/// enough that it will work with any error with a reasonable [`core::fmt::Display`] impl.
+#[macro_export]
+macro_rules! assert_error_matches {
```
The name sounds very generic to me (like it could be used for non-parsing stuff).  
Maybe something like `assert_parse_error_matches!` ?

ggwpez commented on substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs:
```diff
@@ -0,0 +1,259 @@
+	path: P,
+	closure: F,
+) {
+	use std::{env::*, path::*};
```
I think we should define the Mutex here, because this is the only place that we are using it and it is private anyway.

ggwpez commented on substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs:
```diff
@@ -0,0 +1,259 @@
+	use std::{env::*, path::*};
+
+	// avoid race condition when swapping out `CARGO_MANIFEST_DIR`
+	let guard = MANIFEST_DIR_LOCK.lock().unwrap();
```
It should be possible to paralyze this (since we use the same path for all tests anyway), but we can do it later.

sam0x17 replied:
> paralyze

🥲

ggwpez commented on substrate/frame/support/src/tests/mod.rs:
```diff
@@ -77,13 +84,33 @@ pub mod frame_system {
 	pub enum Error<T> {
 		/// Required by construct_runtime
 		CallFiltered,
+		/// Used in tasks example
```
```suggestion
		/// Used in tasks example.
```

ggwpez commented on substrate/frame/support/src/traits/tasks.rs:
```diff
@@ -0,0 +1,82 @@
+//! service work, and supporting types.
+
+/// Contain's re-exports of all the supporting types for the [`Task`] trait. Used in the macro
+/// expansion of `RuntimeTask`.
```
Pure macro exports should be in `__private`.

ggwpez commented on substrate/frame/support/src/traits/tasks.rs:
```diff
@@ -0,0 +1,82 @@
+	pub use sp_weights::Weight;
+}
+
+use prelude::*;
```
This is not good practice, since we pollute everything with SemVer abiding re-exports.

sam0x17 replied:
fair, IIRC this only gets used as the first line of some of the generated functions and never gets used in a user scope, but fair enough

ggwpez commented on substrate/frame/system/src/lib.rs:
```diff
@@ -628,6 +636,28 @@ pub mod pallet {
+
+		#[pallet::call_index(8)]
+		#[pallet::weight(task.weight())]
+		pub fn do_task(origin: OriginFor<T>, task: T::RuntimeTask) -> DispatchResultWithPostInfo {
```
So we are not calling this from anywhere yet, or?

gupnik replied:
Correct - just an extrinsic for now.

ggwpez reviewed (approved):
Nice!

gupnik commented:
bot fmt

ggwpez commented on substrate/frame/support/src/traits/tasks.rs:
```diff
@@ -0,0 +1,87 @@
+use scale_info::TypeInfo;
+use sp_runtime::DispatchError;
+use sp_std::{fmt::Debug, iter::Iterator, vec, vec::IntoIter};
+use sp_weights::Weight;
```
Why are they still exported here?

gupnik replied:
They are being used in the trait definition below (not being exported). The export is separate via `__private`.

gupnik commented:
bot fmt

sam0x17 commented:
🎉

louismerlin reviewed (approved):
Approving for bot, SRLabs will review this in the future, PR is in pipeline

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-6-0/5855/1
